### ISSUE TITLE
ProxylessNas:how to use InputChoice in ProxylessNas algorithm?

### ISSUE THREAD

**Issue by gqq99** [microsoft/nni#2367]
Recently i'm trying to build my superNet and want to search using proxylessNAS algorithm,but i found that it seems not to support use InputChoice in ProxylessNAS algorithm,right? I need to refer darts to modify ProxylessNAS ,right?

My superNet like this:

class ENASLayer(mutables.MutableScope):

    def __init__(self, key, prev_labels, in_filters, out_filters):
        super().__init__(key)
        self.in_filters = in_filters
        self.out_filters = out_filters
        self.mutable = mutables.LayerChoice([
            ConvBranch(in_filters, out_filters, 3, 1, 1, separable=False),
            ConvBranch(in_filters, out_filters, 3, 1, 1, separable=True),
            ConvBranch(in_filters, out_filters, 5, 1, 2, separable=False),
            ConvBranch(in_filters, out_filters, 5, 1, 2, separable=True),
            PoolBranch('avg', in_filters, out_filters, 3, 1, 1),
            PoolBranch('max', in_filters, out_filters, 3, 1, 1)
        ])
        if len(prev_labels) > 0:
            self.skipconnect = mutables.InputChoice(choose_from=prev_labels, n_chosen=None)
        else:
            self.skipconnect = None
        self.batch_norm = nn.BatchNorm2d(out_filters, affine=False)

    def forward(self, prev_layers):
        out = self.mutable(prev_layers[-1])
        if self.skipconnect is not None:
            connection = self.skipconnect(prev_layers[:-1])
            if connection is not None:
                out += connection
        return self.batch_norm(out)

class GeneralNetwork(nn.Module):
    def __init__(self, num_layers=12, out_filters=24, in_channels=3, num_classes=10,
                 dropout_rate=0.0):
        super().__init__()
        self.num_layers = num_layers
        self.num_classes = num_classes
        self.out_filters = out_filters

        self.stem = nn.Sequential(
            nn.Conv2d(in_channels, out_filters, 3, 1, 1, bias=False),
            nn.BatchNorm2d(out_filters)
        )

        pool_distance = self.num_layers // 3
        self.pool_layers_idx = [pool_distance - 1, 2 * pool_distance - 1]
        self.dropout_rate = dropout_rate
        self.dropout = nn.Dropout(self.dropout_rate)

        self.layers = nn.ModuleList()
        self.pool_layers = nn.ModuleList()
        labels = []
        for layer_id in range(self.num_layers):
            labels.append("layer_{}".format(layer_id))
            if layer_id in self.pool_layers_idx:
                self.pool_layers.append(FactorizedReduce(self.out_filters, self.out_filters))
            self.layers.append(ENASLayer(labels[-1], labels[:-1], self.out_filters, self.out_filters))

        self.gap = nn.AdaptiveAvgPool2d(1)
        self.dense = nn.Linear(self.out_filters, self.num_classes)

    def forward(self, x):
        bs = x.size(0)
        cur = self.stem(x)

        layers = [cur]

        for layer_id in range(self.num_layers):
            cur = self.layers[layer_id](layers)
            layers.append(cur)
            if layer_id in self.pool_layers_idx:
                for i, layer in enumerate(layers):
                    layers[i] = self.pool_layers[self.pool_layers_idx.index(layer_id)](layer)
                cur = layers[-1]

        cur = self.gap(cur).view(bs, -1)
        cur = self.dropout(cur)
        logits = self.dense(cur)
        return logits

**nni Environment**:
- nni version:
- nni mode(local|pai|remote):
- OS:
- python version:
- is conda or virtualenv used?: 
- is running in docker?:

**Comment by gqq99** [comment]
My supertNet is as below:     and I want to use proxylessNAS,but  *****\pytorch\base_mutator.py:27:self._structured_mutables = self._parse_search_space(self.model) went wrong!

class GameSRLayer(nn.Module):

    def __init__(self, key, prev_labels, in_filters, out_filters,Skip=True):
        # super().__init__(key)
        print('key is',key)
        super().__init__()
        self.in_filters = in_filters
        self.out_filters = out_filters
        op_condidats = [
            ConvRelu(self.in_filters, self.out_filters, 1,1, padding=0),
            ConvRelu(self.in_filters, self.out_filters, 3,1, padding=1),
            ConvRelu(self.in_filters, self.out_filters, 5,1, padding=2)
        ]
        if in_filters == out_filters:
            op_condidats += [ZeroLayer(stride=1)]
        self.mutable = mutables.LayerChoice(op_condidats,key=key)
        if len(prev_labels) > 1 and Skip:
            print('len(prev_labels) is',len(prev_labels))
            pprint.pprint(prev_labels)
            self.skipconnect = mutables.InputChoice(choose_from=prev_labels[1:], n_chosen=1,key="ip_{}".format(len(prev_labels)))
        else:
            self.skipconnect = None
        

    def forward(self, prev_layers):
        out = self.mutable(prev_layers[-1])
        if self.skipconnect is not None:
            connection = self.skipconnect(prev_layers[:-1])
            if connection is not None:
                out += connection
        return out
    @staticmethod
    def is_zero_layer():
        return False


class GeneralNetwork(nn.Module):
    def __init__(self, num_layers=8, out_filters=24, in_channels=1, scale=2,
                 dropout_rate=0.0):
        super().__init__()
        self.num_layers = num_layers
        self.out_filters = out_filters
        self.scale = scale
        self.in_channels = in_channels
        #def __init__(self, C_in, C_out, kernel_size, stride, padding)

        
        self.stem = nn.Sequential(
            nn.Conv2d(self.in_channels, self.out_filters, 3, 1, 1),
            nn.ReLU()
        )
        
        self.layers = nn.ModuleList()
        
        labels = []

    
        ## middle layer
        for layer_id in range(self.num_layers-1):
            labels.append("layer_{}".format(layer_id))
            
            self.layers.append(GameSRLayer(labels[-1], labels[:-1], self.out_filters, self.out_filters))

        ##last layer
        #
        self.last_layer = nn.Sequential(
            nn.Conv2d(self.out_filters, self.scale*self.scale, 3, 1, 1),
            nn.ReLU()
        )

    
        self.dense = nn.Linear(self.out_filters, self.num_classes)
        

    def forward(self, x):
       
        cur = self.stem(x)

        layers = [cur]

        for layer_id in range(self.num_layers):
            cur = self.layers[layer_id](layers)
            layers.append(cur)

        cur = self.last_layer(cur)
           

        cur = cur.view(bs, -1)
        logits = self.dense(cur)
      
        return logits

**Comment by QuanluZhang** [comment]
@gqq99 yes we have not supported InputChoice in proxylessnas trainer, because there is only layer choice in proxylessnas search space.

**Comment by gqq99** [comment]
@QuanluZhang Thanks! If I want to complete the search for the above network, what nas algorithm do you recommend?

**Comment by QuanluZhang** [comment]
@gqq99 I would suggest you to try ENAS, DARTS and SPOS first. As there are randomness and the weight sharing based approaches are a little unstable, it is hard to say which one would produce the best architecture for your task. It is worth to try them all and check the results.

**Comment by gqq99** [comment]
@QuanluZhang Thanks very much!
